examples/SimilarityLearning/mnist-embeddings.py

```diff
@@ -0,0 +1,315 @@
+
+        # if "x" was a list of tensors, then split the embeddings
+        # note the call has changed in v0.12 "split(value, num_or_size_splits, axis=0, ...)"
+        if list_split > 0:
+            embeddings = tf.split(embeddings, list_split, 0)
+
+        return embeddings
+
+    def _get_input_vars(self):
+        return [InputVar(tf.float32, (None, 28, 28), 'input')]
+
+    def _build_graph(self, input_vars):
+        tf.identity(self.embed(input_vars), name="emb")
+
+
+class MnistPairs(dataset.Mnist):
+    """We could also write
+        ds = dataset.Mnist('train')
+        ds = JoinData([ds, ds])
+        ds = MapData(ds, lambda dp: [dp[0], dp[2], dp[1] == dp[3]])
+        ds = BatchData(ds, 128 // 2)
+
+    but then the positives pairs would be really rare (p=0.1).
+    Note, we do not override the method size() for practical considerations.
+    """
+    def __init__(self, train_or_test, shuffle=True, dir=None):
+        super(MnistPairs, self).__init__(train_or_test, shuffle, dir)
+        # now categorize these digits
+        self.data_dict = []
+        for clazz in range(0, 10):
+            clazz_filter = np.where(self.labels == clazz)
+            self.data_dict.append(self.images[clazz_filter])
+
+    def get_data(self):
+        while True:
+            pick_label = self.rng.randint(10)
+            pick_other = pick_label
+            y = self.rng.randint(2)
+
+            if y == 0:
+                # pair with different digits
+                offset = self.rng.randint(9)
+                pick_other = (pick_label + offset + 1) % 10
+                assert not pick_label == pick_other
+
+            l = self.rng.randint(len(self.data_dict[pick_label]))
+            r = self.rng.randint(len(self.data_dict[pick_other]))
+
+            l = np.reshape(self.data_dict[pick_label][l], [28, 28]).astype(np.float32)
+            r = np.reshape(self.data_dict[pick_other][r], [28, 28]).astype(np.float32)
+
+            yield [l, r, y]
+
+
+class SiameseModel(EmbeddingModel):
+    @staticmethod
+    def get_data():
+        # return [digit_A, digit_B, similar?]
+        # return [digit_A, digit_B, similar?]
+        ds = MnistPairs('train')
+        ds = BatchData(ds, 128 // 2)
+        return ds
+
+    def _get_input_vars(self):
+        return [InputVar(tf.float32, (None, 28, 28), 'input'),
+                InputVar(tf.float32, (None, 28, 28), 'input_y'),
+                InputVar(tf.int32, (None,), 'label')]
+
+    def _build_graph(self, input_vars):
+        # get inputs
+        x, y, label = input_vars
+        # embed them
+        x, y = self.embed([x, y])
+        # tag them, just for inference later on
+        x = tf.identity(x, name="emb")
+
+        # compute the actual loss
+        cost, pos_dist, neg_dist = symbf.contrastive_loss(x, y, label, 5., extra=True)
+        self.cost = tf.identity(cost, name="cost")
+
+        # track these values during training
+        tf.summary.scalar('pos_dist', pos_dist)
```

Comment:
summary op only gets run once after each epoch (and on the last data point only). For scalar summaries you probably want to use `add_moving_summary`, which computes moving average over every iteration.